botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,275 @@
+    def __init__(
+        self,
+        loss_function_hyperparameters: Dict[str, Any] = dict(
+            dist_weight=1.0, dist_threshold=0.5
+        ),
+    ) -> None:
+        r"""Batch loss function for the task of finding top-K.
+
+        Args:
+            loss_function_hyperparameters: hyperparameters for the loss function class.
+        """
+
+        super().__init__()
+        self.register_buffer(
+            "dist_weight", torch.as_tensor(loss_function_hyperparameters["dist_weight"])
+        )
+        self.register_buffer(
+            "dist_threshold",
+            torch.as_tensor(loss_function_hyperparameters["dist_threshold"]),
+        )
+
+    def forward(self, A: Tensor, Y: Tensor) -> Tensor:
+        r"""Evaluate batch loss function on a tensor of actions.
+
+        Args:
+            A: Actor tensor with shape `batch_size x n_fantasy_at_design_pts
+                x num_actions x action_dim`.
+            Y: Fantasized sample with shape `n_fantasy_at_action_pts x
+                n_fantasy_at_design_pts x batch_size x num_actions`.
+
+        Returns:
+            A Tensor of shape `n_fantasy_at_action_pts x batch`.
+        """
+
+        Y = Y.sum(dim=-1).mean(dim=0)
+        # >>> n_fantasy_at_design_pts x batch_size
+
+        num_actions = A.shape[-2]
+
+        dist_reward = 0
+        if num_actions >= 2:
+            A_distance = torch.cdist(A.contiguous(), A.contiguous(), p=1.0)
+            A_distance_triu = torch.triu(A_distance)
+            # >>> n_fantasy_at_design_pts x batch_size x num_actions x num_actions
+
+            A_distance_triu[A_distance_triu > self.dist_threshold] = self.dist_threshold
+            denominator = num_actions * (num_actions - 1) / 2.0
+            dist_reward = A_distance_triu.sum((-1, -2)) / denominator
+            # >>> n_fantasy_at_design_pts x batch_size
+
+        q_hes = Y + self.dist_weight * dist_reward
+        # >>> n_fantasy_at_design_pts x batch_size
+
+        return q_hes
+
+
+class qLossFunctionMinMax(nn.Module):
+    def __init__(
+        self,
+        loss_function_hyperparameters: Optional[Dict[str, Any]] = dict(),
+    ) -> None:
+        r"""Loss function for task of finding min and max.
+
+        Args:
+            loss_function_hyperparameters: hyperparameters for the loss function class
+        """
+        super().__init__()
+
+    def forward(self, A: Tensor, Y: Tensor) -> Tensor:
+        r"""Evaluate batch loss function on a tensor of actions.
+
+        Args:
+            A: Actor tensor with shape `batch_size x n_fantasy_at_design_pts
+                x num_actions x action_dim`.
+            Y: Fantasized sample with shape `n_fantasy_at_action_pts x
+                n_fantasy_at_design_pts x batch_size x num_actions`.
+
+        Returns:
+            A Tensor of shape `n_fantasy_at_action_pts x batch`.
+        """
+
+        assert A.shape[-2] == 2
+
+        Y[:, :, :, 0] = -1 * Y[:, :, :, 0]
+
+        q_hes = Y.sum(dim=-1).mean(dim=0)
```

Comment:
This is problematic as you're changing the input tensor in-place. That is, the `Y` that you passed to `forward` outsize will change once you call the objective on it. How about this?
```suggestion
        q_hes = (Y[..., 1:].sum(dim=-1) - Y[..., 0]).mean(dim=0)
```